Refactor to Next.js and add blog section

charliegerard-stripe opened the pull request:
## Description
This PR fixes some semantic issues and refactors the website to Next.js so we can use Markdoc for future improvements. 
As an MVP, it also adds a "Blog posts" section that fetches the latest posts from dev.to, displays the 3 latest ones and redirects to dev.to to read more.

_The code needs a bigger refactor that will be handled in future PRs._

**Before**

<img width="1789" alt="Screen Shot 2022-09-08 at 12 35 59 PM" src="https://user-images.githubusercontent.com/100887638/189401523-76861f8f-196b-4365-93dc-07c9c0027469.png">

**After**

![Screen Shot 2022-09-09 at 10 04 39 AM](https://user-images.githubusercontent.com/100887638/189404618-5bf0cb6c-9d9c-48f5-9e67-31ad434da4cb.png)

## Test plan

As this website is hosted using GitHub pages, it needs an index.html file at the root of the project. This file is created after running `npm run build`. This command creates an `out` folder with the prod files in it. **If you make changes to this PR, remember to move the prod files outside of the `out` folder before pushing changes to this PR.**

If you want to launch the site locally, you need to:
* Fetch this PR
* Run `python -m http.server 3000` and check `http://localhost:3000` in your browser.

If you want to make changes, you can run `npm run dev` to check the site locally. 

**If you make changes to this PR and want to be sure about what it will look like once published, run the python command and what displays in the browser should be what will be published once the PR is merged.** 

Hopefully, we'll update this workflow if we can move away from GitHub pages.

charliegerard-stripe committed:
Refactor to Next.js and add blog section

charliegerard-stripe committed:
update padding

mikeb-stripe commented on package.json:
```diff
@@ -0,0 +1,22 @@
+{
+  "name": "stripe-dev-revamp",
```
I think we'll want to re-rename this - stripe-dev should do 🙌🏻

mikeb-stripe commented on components/ClientLibraries.js:
```diff
@@ -0,0 +1,65 @@
+      <section className="client-libs">
+        <a href="https://github.com/stripe/stripe-ruby" target="_blank">
+          <h3>
+            <img src="/images/ruby-logo.svg" />
```
Not sure where the best place is for this comment - I pulled down this branch and none of these images are working. I think they need to be moved to `/public/images/*`

mikeb-stripe replied:
This UI is also quite tall on mobile devices - can we use a flex or a grid to display these?

<img width="241" alt="image" src="https://user-images.githubusercontent.com/102179546/189414832-eec6cf8c-960a-41a8-b54b-01ed596803cd.png">

charliegerard-stripe replied:
What do you think about that?

![Screen Shot 2022-09-09 at 11 36 48 AM](https://user-images.githubusercontent.com/100887638/189420782-e73c96cb-ed86-4f00-93d8-741e413b8bed.png)

mikeb-stripe replied:
Love it!

mikeb-stripe commented on pages/index.js:
```diff
@@ -0,0 +1,357 @@
+    <div className={styles.container}>
+      <Head>
+        <title>Stripe Developer Resources</title>
+        <link
```
I'm getting a console error about this stylesheet - I think it should probably be moved to `_app.js` -

<img width="1215" alt="image" src="https://user-images.githubusercontent.com/102179546/189415297-f798d052-ee72-4be3-b84e-4575a7b464f2.png">

charliegerard-stripe replied:
I realized we don't even use that font anywhere so i'm removing it.

mikeb-stripe commented on pages/index.js:
```diff
@@ -0,0 +1,357 @@
+          rel="stylesheet"
+        />
+        <link rel="shortcut icon" href="favicon.ico" />
+        <meta charset="utf-8" />
```
one more console error, apparently the React-y way to do this is `charSet="utf-8"`

<img width="649" alt="image" src="https://user-images.githubusercontent.com/102179546/189415366-db0e6120-7b7f-44df-964e-208ab136db80.png">

mikeb-stripe commented on components/ClientLibraries.js:
```diff
@@ -0,0 +1,65 @@
+        </a>
+        <a href="https://github.com/stripe/stripe-node" target="_blank">
+          <h3>
+            <img src="/images/node-logo.svg" />
```
In order to be accessible, these img tags should have alt set  - `alt="Node"`, etc

mikeb-stripe commented on pages/index.js:
```diff
@@ -0,0 +1,357 @@
+          href="https://fonts.googleapis.com/css?family=Source+Code+Pro"
+          rel="stylesheet"
+        />
+        <link rel="shortcut icon" href="favicon.ico" />
```
Favicon isn't rendering for me, I _think_ that will also be solved when images are moved to `/public/`

charliegerard-stripe replied:
The issue with the images is that running the app with `npm run dev` needs to find the images in the `public` folder but once built, they need to be at the root of the project, directly in the `images` folder :/ so for now i duplicated the folder until we can move away from GitHub pages. If we host this site on Netlify or Vercel, it will be a lot easier.

mikeb-stripe replied:
Hm - could we create a symlink from one to the other? 

```bash
ln -s /images /public/images
```

Should tie them together, I think? I forget which order we'd want to use, but my gut is to leave the images in `/public/images` so it all works in dev mode

charliegerard-stripe replied:
This is something we'd have to run locally right? Or do you mean adding it somewhere in the project?

mikeb-stripe reviewed (changes requested):
This is looking really good! I'm excited to have this set up. Thanks for taking the initiative, and putting in the work to do this, @charliegerard-stripe.

A couple little nits and minor bugs pointed out in comments, feel free to ping me if I can clarify anything.

mikeb-stripe commented:
@charliegerard-stripe  one more thing - can you set up eslint with `npm run lint` or `yarn lint`, and see what else that brings up? I think Next.js recommends their strict ruleset as the default these days. That ought to be a good place to start.

charliegerard-stripe committed:
start implementing feedback

charliegerard-stripe committed:
implement feedback

charliegerard-stripe committed:
implement feedback

charliegerard-stripe committed:
implement feedback

charliegerard-stripe committed:
implement feedback

charliegerard-stripe commented:
@mikeb-stripe I implemented most of the feedback, let me know what you think! 

I fixed most of the warnings from `npm run lint` except the ones about using their built-in `Image` component instead of the native `img` tag. I think we should probably do in the future but for now, their component breaks the layout as they deal with width and height differently than the styles we're currently using 🤔

mikeb-stripe commented on pages/index.js:
```diff
@@ -0,0 +1,395 @@
+          <div className="sign-up">
+            <p>
+              Stay up to date with the latest from Stripe’s API and developer
+              products:
```
I think this may have been my bug - this needs a space

```suggestion
              products:{' '}
```

mikeb-stripe reviewed (changes requested):
Nice! Thanks for the updates. 

I just noticed in the files window of the PR on GitHub that the `_next/static` directory is included in the repo - should we add that to `.gitignore`, and remove those files from this PR? I don't think they need to be in source control

charliegerard-stripe commented:
It looks like the `index.html` file uses assets from the static folder so if I add it to the `.gitignore` file, there won't be any style on the page :/ 
It should really only be an issue because of GitHub pages so we can definitely remove all that later

charliegerard-stripe committed:
implement feedback

charliegerard-stripe committed:
implement feedback

mikeb-stripe reviewed (approved):
Thanks @charliegerard-stripe - it might be a good idea to document the things we want to change here in the future, if github pages is no longer the host we choose.

Given everything you've said - everything looks good from here 👍
